core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala

```diff
@@ -489,6 +491,29 @@ class StandaloneDynamicAllocationSuite
     }
   }
 
+  test("executor registration on a blacklisted host must fail") {
+    sc = new SparkContext(appConf.set(config.BLACKLIST_ENABLED.key, "true"))
+    val endpointRef = mock(classOf[RpcEndpointRef])
+    val mockAddress = mock(classOf[RpcAddress])
+    when(endpointRef.address).thenReturn(mockAddress)
+    val message = RegisterExecutor("one", endpointRef, "localhost", 10, Map.empty)
+
+    // Get "localhost" on a blacklist.
+    val taskScheduler = mock(classOf[TaskSchedulerImpl])
+    when(taskScheduler.nodeBlacklist()).thenReturn(Set("localhost"))
+    when(taskScheduler.sc).thenReturn(sc)
+    sc.taskScheduler = taskScheduler
+
+    // Create a fresh scheduler backend to blacklist "localhost".
+    sc.schedulerBackend.stop()
+    val backend =
+     new StandaloneSchedulerBackend(taskScheduler, sc, Array(masterRpcEnv.address.toSparkURL))
+    backend.start()
+
+    backend.driverEndpoint.ask[Boolean](message)
+    verify(endpointRef).send(RegisterExecutorFailed(any()))
```

Comment:
You need to wrap this in an `eventually` since its async.

btw, a simple way to see if your test is flaky locally is to run it in in a loop:

```scala
  (0 until 100).foreach { idx =>
    test(s"executor registration on a blacklisted host must fail $idx") {
```

and then in sbt

```
test-only *StandaloneDynamicAllocationSuite -- -z "must fail"
```